corehq/apps/accounting/tests/test_autopay_with_api.py

```diff
@@ -115,25 +116,31 @@ def test_get_autopayable_invoices_returns_nothing(self):
         autopayable_invoices = Invoice.autopayable_invoices(date_due)
         self.assertCountEqual(autopayable_invoices, [])
 
-    def test_pay_autopayable_invoices(self):
+    # Keys for idempotent requests can only be used with the same parameters they were first used with.
+    # So introduce randomness in idempotency key to avoid clashes
+    @patch('corehq.apps.accounting.models.Invoice.invoice_number', new_callable=lambda: str(uuid.uuid4()))
+    def test_pay_autopayable_invoices(self, fake_invoice_id):
```

Comment:
This appears to be to be patching a property with a lambda function:

https://github.com/dimagi/commcare-hq/blob/0a14f565bb5c51211733da2e8a0ad60e7557f70e/corehq/apps/accounting/models.py#L2095-L2096
Which means that test code that accesses `some_invoice.invoice_number` will get the lambda function and not a UUID string. Interesting that that appears to have worked.

Would it make more sense to use [`PropertyMock`](https://docs.python.org/3/library/unittest.mock.html#unittest.mock.PropertyMock)?
```suggestion
    @patch('corehq.apps.accounting.models.Invoice.invoice_number', new_callable=PropertyMock)
    def test_pay_autopayable_invoices(self, fake_invoice_id):
        fake_invoice_id.return_value = str(uuid.uuid4())
```

Don't forget to import `PropertyMock` at the top of this file if you accept the suggestion.

A simpler option that will probably also work would be to replace the property with a UUID string:
```suggestion
    @patch('corehq.apps.accounting.models.Invoice.invoice_number', str(uuid.uuid4()))
    def test_pay_autopayable_invoices(self, fake_invoice_id):
```

With any of these options all invoices will have the same invoice number while the mock is in place. I assume that's not a problem, but may be worth considering.

The same applies to the test below as well.

Reply:
Hi Daniel, I tried patching the `invoice_number` with a return value like `@patch('....invoice_number', return_value=str(uuid.uuid4()))`, and when `some_invoice.invoice_number` is accessed like an attribute, a MagicMock object is returned, only when `some_invoice.invoice_number()` is called, I can get the string... I don't understand why, but change it to `@patch('....invoice_number', new_callable=lambda:....)` seems like giving me the result I want.
I tried your first approach and think this is more readable than a lambda function. Fixed here: 168ef78f17d4d99b5fccc6579f5642fee01c2420

Reply:
>  I tried patching the `invoice_number` with a return value like `@patch('....invoice_number', return_value=str(uuid.uuid4()))`

The problem is `return_value=`, which makes the mock a callable. I think removing `return_value=` should do what you want (see my second suggestion above).

`PropertyMock` works too, it's just a little more verbose.

Reply:
Hi Daniel, the second approach would gave `TypeError: test_pay_autopayable_invoices() missing 1 required positional argument: 'fake_invoice_id'` ... In most examples, when we patch the method or attribute, we will specify the behavior, like `return_value`, `side_effect` ... Without specifying the behavior, what're we patching? I'm wondering, does this error mean, we're not passing any mock object to the test function?
After removing `fake_invoice_id` from the test function arguments: `def test_pay_autopayable_invoices(self):`, it works. Fixed in 2c543a291589823093be4f703891341e4daed3b8